sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala

```diff
@@ -68,10 +72,57 @@ object InferWindowGroupLimit extends Rule[LogicalPlan] with PredicateHelper {
     case _ => false
   }
 
+  private def limitSupport(limit: Int, window: Window): Boolean =
+    limit <= conf.windowGroupLimitThreshold && window.child.maxRows.forall(_ > limit) &&
+      !window.child.isInstanceOf[WindowGroupLimit] &&
+      window.windowExpressions.forall(isExpandingWindow) &&
+      window.orderSpec.exists(!_.foldable) &&
+      // LimitPushDownThroughWindow have better performance than WindowGroupLimit if the
+      // window function is Rank, DenseRank and RowNumber, and Window partitionSpec is empty.
+      !(window.partitionSpec.isEmpty && supportsPushdownThroughWindow(window.windowExpressions) &&
```

Comment:
It seems `LimitPushDownThroughWindow` already pushdown `Limit` through `Window` before arrive `InferWindowGroupLimit`.

Reply:
yea, I will remove it.

Reply:
No, after `LimitPushDownThroughWindow` is pushed down limit, the new plan will still hit `InferWindowGroupLimit`.

Reply:
Could you give an example?

Reply:
`df.withColumn("rn", row_number().over(Window.orderBy($"order".asc_nulls_first)))
              .limit(10)
              .where("rn <= 5")`
 Because filter in `InferWindowGroupLimit` is possible to convert filter to limit. Code is 
```
case (limit, rankLikeFunction) if limit <= conf.windowGroupLimitThreshold &&
              child.maxRows.forall(_ > limit) =>
              if (limit > 0) {
                val newFilterChild = if (rankLikeFunction.isInstanceOf[RowNumber] &&
                  partitionSpec.isEmpty && limit < conf.topKSortFallbackThreshold) {
                  // Top n (Limit + Sort) have better performance than WindowGroupLimit if the
                  // window function is RowNumber and Window partitionSpec is empty.
                  Limit(Literal(limit), window)
```

Reply:
InferWindowGroupLimitSuite.test("Insert window group limit node for top-k computation: empty relation") tested.

Reply:
`InferWindowGroupLimitSuite` not contains `LimitPushDownThroughWindow` before `InferWindowGroupLimit`.
Could you test with `DataFrameWindowFunctionsSuite`?

Reply:
Done.